File: editors/vscode/src/mergeOwnershipStateChart.ts
import {
    ChartBuilder,
    StateBuilder,
    StateChart,
} from '@lazily-hub/lazily-js/statechart';

/**
 * VS Code-side projection of the binary-owned merge-ownership state machine.
 *
 * The controller remains authoritative for disk-write permission. JavaScript
 * has no shared-thread reactive Context, so this chart owns only the extension
 * host's local replica lifecycle and uses the same phase/event vocabulary as
 * `agent_doc_merge::ownership`.
 */
export const MERGE_OWNERSHIP_PHASES = [
    'detached',
    'attached',
    'editor_owns_buffer',
    'binary_write_requested',
    'lazily_patch_applied_proven',
    'committed',
] as const;

export type MergeOwnershipPhase = (typeof MERGE_OWNERSHIP_PHASES)[number];

export const MERGE_OWNERSHIP_EVENTS = [
    'editor_attached',
    'editor_buffer_observed',
    'editor_detached',
    'heartbeat_stale',
    'binary_write_requested',
    'lazily_patch_applied_observed',
    'committed',
] as const;

export type MergeOwnershipEvent = (typeof MERGE_OWNERSHIP_EVENTS)[number];

const ROOT = 'merge_ownership';

const DEFINITION = new ChartBuilder()
    .state(StateBuilder.compound(ROOT, 'detached'))
    .state(
        StateBuilder.atomic('detached')
            .parent(ROOT)
            .on('editor_attached', 'attached')
            .on('editor_buffer_observed', 'editor_owns_buffer')
            .on('editor_detached', 'detached')
            .on('committed', 'committed'),
    )
    .state(
        StateBuilder.atomic('attached')
            .parent(ROOT)
            .on('editor_attached', 'attached')
            .on('editor_buffer_observed', 'editor_owns_buffer')
            .on('editor_detached', 'detached')
            .on('heartbeat_stale', 'detached'),
    )
    .state(
        StateBuilder.atomic('editor_owns_buffer')
            .parent(ROOT)
            .on('editor_attached', 'attached')
            .on('editor_buffer_observed', 'editor_owns_buffer')
            .on('editor_detached', 'detached')
            .on('binary_write_requested', 'binary_write_requested'),
    )
    .state(
        StateBuilder.atomic('binary_write_requested')
            .parent(ROOT)
            .on('binary_write_requested', 'binary_write_requested')
            .on('lazily_patch_applied_observed', 'lazily_patch_applied_proven'),
    )
    .state(
        StateBuilder.atomic('lazily_patch_applied_proven')
            .parent(ROOT)
            .on('lazily_patch_applied_observed', 'lazily_patch_applied_proven')
            .on('committed', 'committed'),
    )
    .state(
        StateBuilder.final('committed')
            .parent(ROOT)
            .on('committed', 'committed'),
    )
    .build();

export class MergeOwnershipStateChart {
    private readonly chart = new StateChart(DEFINITION);

    get phase(): MergeOwnershipPhase {
        const activeLeaf = this.chart.activeLeaves()[0];
        if (!MERGE_OWNERSHIP_PHASES.includes(activeLeaf as MergeOwnershipPhase)) {
            throw new Error(`unknown merge-ownership phase: ${activeLeaf ?? '<none>'}`);
        }
        return activeLeaf as MergeOwnershipPhase;
    }

    get editorAttached(): boolean {
        const phase = this.phase;
        return phase !== 'detached' && phase !== 'committed';
    }

    send(event: MergeOwnershipEvent): boolean {
        return this.chart.send(event);
    }
}

File: editors/jetbrains/src/main/kotlin/com/github/btakita/agentdoc/MergeOwnershipStateChart.kt
package com.github.btakita.agentdoc

import io.github.lazily.ChartBuilder
import io.github.lazily.StateBuilder
import io.github.lazily.ThreadSafeContext
import io.github.lazily.ThreadSafeStateChart

/**
 * JetBrains-side projection of the binary-owned merge-ownership state machine.
 *
 * The controller remains authoritative for disk-write permission. This chart
 * gives the plugin one thread-safe lifecycle fact instead of an independently
 * mutated `attached` flag while preserving the transition vocabulary used by
 * `agent_doc_merge::ownership`.
 */
internal enum class MergeOwnershipPhase(val stateId: String) {
    Detached("detached"),
    Attached("attached"),
    EditorOwnsBuffer("editor_owns_buffer"),
    BinaryWriteRequested("binary_write_requested"),
    LazilyPatchAppliedProven("lazily_patch_applied_proven"),
    Committed("committed"),
}

internal enum class MergeOwnershipEvent(val eventId: String) {
    EditorAttached("editor_attached"),
    EditorBufferObserved("editor_buffer_observed"),
    EditorDetached("editor_detached"),
    HeartbeatStale("heartbeat_stale"),
    BinaryWriteRequested("binary_write_requested"),
    LazilyPatchAppliedObserved("lazily_patch_applied_observed"),
    Committed("committed"),
}

internal class MergeOwnershipStateChart(context: ThreadSafeContext) {
    private val chart = ThreadSafeStateChart(DEFINITION, context)

    val phase: MergeOwnershipPhase
        get() {
            val activeLeaf = chart.activeLeaves().single()
            return MergeOwnershipPhase.entries.single { it.stateId == activeLeaf }
        }

    val editorAttached: Boolean
        get() =
            when (phase) {
                MergeOwnershipPhase.Detached,
                MergeOwnershipPhase.Committed,
                -> false

                MergeOwnershipPhase.Attached,
                MergeOwnershipPhase.EditorOwnsBuffer,
                MergeOwnershipPhase.BinaryWriteRequested,
                MergeOwnershipPhase.LazilyPatchAppliedProven,
                -> true
            }

    fun send(event: MergeOwnershipEvent): Boolean = chart.send(event.eventId)

    private companion object {
        const val ROOT = "merge_ownership"

        val DEFINITION =
            ChartBuilder()
                .state(StateBuilder.compound(ROOT, MergeOwnershipPhase.Detached.stateId))
                .state(
                    StateBuilder.atomic(MergeOwnershipPhase.Detached.stateId)
                        .parent(ROOT)
                        .on(
                            MergeOwnershipEvent.EditorAttached.eventId,
                            MergeOwnershipPhase.Attached.stateId,
                        )
                        .on(
                            MergeOwnershipEvent.EditorBufferObserved.eventId,
                            MergeOwnershipPhase.EditorOwnsBuffer.stateId,
                        )
                        .on(
                            MergeOwnershipEvent.EditorDetached.eventId,
                            MergeOwnershipPhase.Detached.stateId,
                        )
                        .on(
                            MergeOwnershipEvent.Committed.eventId,
                            MergeOwnershipPhase.Committed.stateId,
                        ),
                )
                .state(
                    StateBuilder.atomic(MergeOwnershipPhase.Attached.stateId)
                        .parent(ROOT)
                        .on(
                            MergeOwnershipEvent.EditorAttached.eventId,
                            MergeOwnershipPhase.Attached.stateId,
                        )
                        .on(
                            MergeOwnershipEvent.EditorBufferObserved.eventId,
                            MergeOwnershipPhase.EditorOwnsBuffer.stateId,
                        )
                        .on(
                            MergeOwnershipEvent.EditorDetached.eventId,
                            MergeOwnershipPhase.Detached.stateId,
                        )
                        .on(
                            MergeOwnershipEvent.HeartbeatStale.eventId,
                            MergeOwnershipPhase.Detached.stateId,
                        ),
                )
                .state(
                    StateBuilder.atomic(MergeOwnershipPhase.EditorOwnsBuffer.stateId)
                        .parent(ROOT)
                        .on(
                            MergeOwnershipEvent.EditorAttached.eventId,
                            MergeOwnershipPhase.Attached.stateId,
                        )
                        .on(
                            MergeOwnershipEvent.EditorBufferObserved.eventId,
                            MergeOwnershipPhase.EditorOwnsBuffer.stateId,
                        )
                        .on(
                            MergeOwnershipEvent.EditorDetached.eventId,
                            MergeOwnershipPhase.Detached.stateId,
                        )
                        .on(
                            MergeOwnershipEvent.BinaryWriteRequested.eventId,
                            MergeOwnershipPhase.BinaryWriteRequested.stateId,
                        ),
                )
                .state(
                    StateBuilder.atomic(MergeOwnershipPhase.BinaryWriteRequested.stateId)
                        .parent(ROOT)
                        .on(
                            MergeOwnershipEvent.BinaryWriteRequested.eventId,
                            MergeOwnershipPhase.BinaryWriteRequested.stateId,
                        )
                        .on(
                            MergeOwnershipEvent.LazilyPatchAppliedObserved.eventId,
                            MergeOwnershipPhase.LazilyPatchAppliedProven.stateId,
                        ),
                )
                .state(
                    StateBuilder.atomic(MergeOwnershipPhase.LazilyPatchAppliedProven.stateId)
                        .parent(ROOT)
                        .on(
                            MergeOwnershipEvent.LazilyPatchAppliedObserved.eventId,
                            MergeOwnershipPhase.LazilyPatchAppliedProven.stateId,
                        )
                        .on(
                            MergeOwnershipEvent.Committed.eventId,
                            MergeOwnershipPhase.Committed.stateId,
                        ),
                )
                .state(
                    StateBuilder.final(MergeOwnershipPhase.Committed.stateId)
                        .parent(ROOT)
                        .on(
                            MergeOwnershipEvent.Committed.eventId,
                            MergeOwnershipPhase.Committed.stateId,
                        ),
                )
                .build()
    }
}
